### PR TITLE
console: show session argument input for computed fields

### LINKED ISSUE #5154: console: session_argument can be set for computed fields as well
Since https://github.com/hasura/graphql-engine/pull/4486 (released in v1.3), even computed fields can supply session_variable argument. This could be shown in UI as well.

Note the server bug in https://github.com/hasura/graphql-engine/issues/3757 . Tldr: a function which is both added as a custom function (via `track_function`) and computed field (via `add_computed_field`) currently breaks introspection. Without fixing this first, ppl might accidently track a computed field (or vice versa) and face a broken schema !

### PR DESCRIPTION
**Pull request by kolharsam**
closes #5154 
### Description

This PR mainly adds a new input field for including a session argument for computed fields. This PR also includes minor fixes and also TS migration of the `ComputedFieldsEditor` file

### Changelog

- [x] `CHANGELOG.md` is updated with user-facing content relevant to this PR. If no changelog is required, then add the `no-changelog-required` label.

### Affected components

- [x] Console

### Function that you use to test this PR

```SQL
CREATE OR REPLACE FUNCTION public.get_session_rolezz(result_row text_result, hasura_session json)
 RETURNS SETOF text_result
 LANGUAGE sql
 STABLE
AS $function$
    (SELECT q.* FROM (VALUES (hasura_session ->> 'x-hasura-role')) q)
    UNION ALL
    (SELECT * from text_result)
$function$
```

Screenshot 2020-11-11 at 19 57 12

### CODE REVIEW

**Review by soorajshankar — APPROVED** [review]
The code looks good, left a minor commend but **Do as you wish**


Observations and steps that I followed to test 
---
1. Run and track the following SQL

```sql
CREATE TABLE public.article (
    id integer NOT NULL,
    book text NOT NULL,
    author integer NOT NULL
);

CREATE TABLE public.article_likes (
    user_id integer NOT NULL,
    article_id integer NOT NULL,
    id integer NOT NULL
);

CREATE TABLE public.author (
    first_name text NOT NULL,
    last_name text NOT NULL,
    id integer NOT NULL
);
CREATE SEQUENCE public.article_id_seq
    AS integer
    START WITH 1
    INCREMENT BY 1
    NO MINVALUE
    NO MAXVALUE
    CACHE 1;
ALTER SEQUENCE public.article_id_seq OWNED BY public.article.id;

CREATE SEQUENCE public.article_likes_id_seq
    AS integer
    START WITH 1
    INCREMENT BY 1
    NO MINVALUE
    NO MAXVALUE
    CACHE 1;
ALTER SEQUENCE public.article_likes_id_seq OWNED BY public.article_likes.id;
CREATE SEQUENCE public.author_id_seq
    AS integer
    START WITH 1
    INCREMENT BY 1
    NO MINVALUE
    NO MAXVALUE
    CACHE 1;
ALTER SEQUENCE public.author_id_seq OWNED BY public.author.id;
ALTER TABLE ONLY public.article ALTER COLUMN id SET DEFAULT nextval('public.article_id_seq'::regclass);
ALTER TABLE ONLY public.article_likes ALTER COLUMN id SET DEFAULT nextval('public.article_likes_id_seq'::regclass);
ALTER TABLE ONLY public.author ALTER COLUMN id SET DEFAULT nextval('public.author_id_seq'::regclass);
ALTER TABLE ONLY public.article_likes
    ADD CONSTRAINT article_likes_pkey PRIMARY KEY (id);
ALTER TABLE ONLY public.article
    ADD CONSTRAINT article_pkey PRIMARY KEY (id);
ALTER TABLE ONLY public.author
    ADD CONSTRAINT author_pkey PRIMARY KEY (id);
ALTER TABLE ONLY public.article
    ADD CONSTRAINT article_author_fkey FOREIGN KEY (author) REFERENCES public.author(id) ON UPDATE RESTRICT ON DELETE RESTRICT;
ALTER TABLE ONLY public.article_likes
    ADD CONSTRAINT article_likes_article_id_fkey FOREIGN KEY (article_id) REFERENCES public.article(id) ON UPDATE RESTRICT ON DELETE RESTRICT;
ALTER TABLE ONLY public.article_likes
    ADD CONSTRAINT article_likes_user_id_fkey FOREIGN KEY (user_id) REFERENCES public.author(id) ON UPDATE RESTRICT ON DELETE RESTRICT;


INSERT INTO "public"."author" ("first_name", "last_name") VALUES ('sooraj', 'sanker');
INSERT INTO "public"."article" ("book", "author") VALUES ('test book', '1');
INSERT INTO "public"."article_likes" ("user_id", "article_id") VALUES ('1', '1');
```

2. Run the following without tracking it

```sql
CREATE FUNCTION public.article_liked_by_user(article_row public.article, hasura_session json) RETURNS boolean
    LANGUAGE sql STABLE
    AS $$
SELECT EXISTS (
    SELECT 1
    FROM article_likes A
    WHERE A.user_id :: text= hasura_session ->> 'x-hasura-user-id' AND A.article_id = article_row.id
);
$$;
```


Add computed field on `articles` table, name it `article_liked_by_user`, add `hasura_session` as session argument

Test the computed field from the gql explorer by passing `x-hasura-user-id = 1`
expected o/p
image


> Note: Functions added as computed fields are filtered from the `trackable_functions` list, this can be tested with a function that returns `SETOF<table>`

File: CHANGELOG.md [hunk 1]
```diff
@@ -85,8 +85,10 @@ This release contains the [PDV refactor (#4111)](https://github.com/hasura/graph
 - cli: allow seeds as alias for seed command (#5693)
 - console: remove ONLY as default for ALTER TABLE in column alter operations (close #5512) #5706
 - console: add option to flag an insertion as a migration from `Data` section (close #1766) (#4933)
+- console: add `<3 hasura` section to view updates and notifications from Hasura (#5070)
```
Comment (beerose) [inline CHANGELOG.md:88]:
```suggestion
```

File: console/src/components/Services/ApiExplorer/OneGraphExplorer/OneGraphExplorer.js [hunk 2]
```diff
@@ -157,8 +157,13 @@ class OneGraphExplorer extends React.Component {
         });
       })
       .catch(err => {
+        console.error(err, err.message);
```
Comment (beerose) [inline console/src/components/Services/ApiExplorer/OneGraphExplorer/OneGraphExplorer.js:160]:
```suggestion
```

File: console/src/components/Services/Data/TableModify/ComputedFieldsEditor.tsx [hunk 3]
```diff
@@ -340,19 +402,32 @@ const ComputedFieldsEditor = ({
...
+              value={computedFieldTableRowArg ?? undefined}
+              placeholder="default: first argument"
               onChange={handleTableRowArgChange}
               className={`form-control ${styles.wd50percent}`}
             />
           </div>
+          <div className={`${styles.add_mar_top}`}>
+            <div className={`${styles.add_mar_bottom_mid}`}>
+              <b>Session argument:</b>
+              <Tooltip message="the function argument into which hasura session variables will be passed" />
```
Comment (beerose) [inline console/src/components/Services/Data/TableModify/ComputedFieldsEditor.tsx:414]:
```suggestion
              <Tooltip message="The function argument into which Hasura session variables will be passed" />
```

File: console/src/components/Services/Data/TableModify/ComputedFieldsEditor.tsx [hunk 4]
```diff
@@ -204,9 +227,18 @@ const ComputedFieldsEditor = ({
...
+        selectedOption: ValueType<OptionTypeBase> | string
+      ) => {
+        // fetch schema fn
+
+        if (!selectedOption) {
+          return;
+        }
+
+        if (selectedOption) {
+          dispatch(fetchFunctionInit((selectedOption as OptionTypeBase).value));
```
Comment (beerose) [inline console/src/components/Services/Data/TableModify/ComputedFieldsEditor.tsx:240]: Can you get rid of type assertion and do a check for `value` field here?

File: console/src/components/Services/Data/TableModify/ComputedFieldsEditor.tsx [hunk 5]
```diff
@@ -216,32 +248,38 @@ const ComputedFieldsEditor = ({
             ...newState[i].definition,
             function: {
               ...newState[i].definition.function,
-              schema: selectedOption.value,
+              schema: (selectedOption as OptionTypeBase).value,
```
Comment (beerose) [inline console/src/components/Services/Data/TableModify/ComputedFieldsEditor.tsx:251]: Same

File: console/src/components/Services/Data/TableModify/ComputedFieldsEditor.tsx [hunk 6]
```diff
@@ -216,32 +248,38 @@ const ComputedFieldsEditor = ({
...
+        }
+
         newState[i] = {
           ...newState[i],
           definition: {
             ...newState[i].definition,
             function: {
               ...newState[i].definition.function,
-              name: selectedOption.value,
+              name: (selectedOption as OptionTypeBase).value,
```
Comment (beerose) [inline console/src/components/Services/Data/TableModify/ComputedFieldsEditor.tsx:274]: Same

File: console/src/components/Services/Data/TableModify/ComputedFieldsEditor.tsx [hunk 7]
```diff
@@ -252,20 +290,47 @@ const ComputedFieldsEditor = ({
...
 
+      const schemaListOptions = schemaList.map(s => getSchemaName(s));
+      const trackedFunctionNames = trackedFunctions.map(
+        func => func.function_name
+      );
+      const functionNameOptions = getSchemaFunctions(
+        functions,
+        computedFieldFunctionSchema
+      )
+        .filter(fn => !trackedFunctionNames.includes(fn.function_name))
```
Comment (beerose) [inline console/src/components/Services/Data/TableModify/ComputedFieldsEditor.tsx:331]:
```suggestion
```

File: console/src/components/Services/Data/TableModify/ComputedFieldsEditor.tsx [hunk 8]
```diff
@@ -64,25 +83,29 @@ const ComputedFieldsEditor = ({
...
-      origComputedFieldComment = origComputedField.comment;
+      if (origComputedField?.comment) {
+        origComputedFieldComment = origComputedField.comment;
+      }
     }
 
     const computedFieldFunctionDef = getQualifiedTableDef(
       computedField.definition.function
-    );
+    ) as TableDefinition;
```
Comment (beerose) [inline console/src/components/Services/Data/TableModify/ComputedFieldsEditor.tsx:97]: Can you get rid of type assertions?

File: console/src/components/Services/Data/TableModify/ComputedFieldsEditor.tsx [hunk 9]
```diff
@@ -64,25 +83,29 @@ const ComputedFieldsEditor = ({
     if (origComputedField) {
       const origComputedFieldFunctionDef = getQualifiedTableDef(
         origComputedField.definition.function
-      );
+      ) as TableDefinition;
```
Comment (beerose) [inline console/src/components/Services/Data/TableModify/ComputedFieldsEditor.tsx:86]: Can you get rid of type assertions?

File: console/src/components/Services/Data/Schema/Schema.js [hunk 10]
```diff
@@ -220,7 +221,25 @@ class Schema extends Component {
...
+        )
+        .flat();
+
+      if (!allComputedFieldFunctions.length) {
+        return allTrackableFunctions;
+      }
+
+      return allTrackableFunctions.filter(
+        func => !allComputedFieldFunctions.includes(func.function_name)
+      );
```
Comment (beerose) [inline console/src/components/Services/Data/Schema/Schema.js:242]: leaving it as a reminder 
```suggestion
```

File: console/src/components/Services/Data/TableModify/ModifyTable.js [hunk 11]
```diff
@@ -356,6 +357,7 @@ const mapStateToProps = (state, ownProps) => ({
   schemaList: state.tables.schemaList,
   postgresVersion: state.main.postgresVersion,
   ...state.tables.modify,
+  trackedFunctions: state.tables.trackedFunctions,
```
Comment (beerose) [inline console/src/components/Services/Data/TableModify/ModifyTable.js:360]:
```suggestion
```

**Review by beerose — CHANGES_REQUESTED** [review]
Left a few, mostly minor, comments.

The functionality looks good.

File: console/src/components/Services/Data/TableModify/ComputedFieldsEditor.tsx [hunk 12]
```diff
@@ -62,27 +77,28 @@ const ComputedFieldsEditor = ({
...
-      );
+      const origComputedFieldFunctionDef =
+        origComputedField.definition.function;
 
       origComputedFieldName = origComputedField.computed_field_name;
       origComputedFieldFunctionName = origComputedFieldFunctionDef.name;
-      origComputedFieldComment = origComputedField.comment;
+      if (origComputedField?.comment) {
+        origComputedFieldComment = origComputedField.comment;
+      }
```
Comment (beerose) [inline console/src/components/Services/Data/TableModify/ComputedFieldsEditor.tsx:87]: A user might have removed a comment. In this case, `origComputedField.comment` would be falsy. As if condition would be `false` for it, then we won't be able to remove a comment.

**Review by beerose — APPROVED** [review]
Console

**Pull request closed without merging**